patterns/unsafe-mods.md

```diff
@@ -0,0 +1,24 @@
+# Contain Unsafety in Small Modules
+
+## Description
+
+If you have `unsafe` code, create the smallest possible module that can uphold the needed invariants to build a minimal safe interface upon the unsafety. Embed this into a larger module that contains only safe code and presents an ergonomic interface. Note that the outer module can contain unsafe functions and methods that call directly into the unsafe code. Users may use this to gain speed benefits.
+
+## Advantages
+
+* This restricts the unsafe code that must be audited
+* Writing the outer module is much easier, since you can count on the guarantees of the inner module
+
+## Disadvantages
+
+* Sometimes, it may be hard to find a suitable interface.
+* The abstraction may introduce inefficiencies.
+
+## Examples
+
+* The [`optional`](https://crates.io/crates/optional) crate has one unsafe module for converting a `&T` to either a slice of one `[T]` or an empty slice.
```

Comment:
There's already a [standard library function](https://doc.rust-lang.org/std/slice/fn.from_ref.html) for doing that, so I suggest using another example :stuck_out_tongue:

Reply:
Yeah, I wrote that before the function became available. The newest version has no more `unsafe`.

I'll look around for other examples.

Reply:
Found one – toolshed keeps all unsafety in submodules.